### PR TITLE
Cluster grenades for uplink and security

### PR DESCRIPTION
**Pull request by Arendian**
## About the PR

Revival of #18348.
This PR refactors the clustergrenade system, adds a multitude of cluster grenade options to the syndicate uplink, puts the clusterbang in the sectech and adds a new grenade to the sectech.

Grenade types:

Security:

- ClusterBang: Splits into three flashbangs after a 3.5 second timer, these flashbangs explode one second after splitting up.
- Stinger grenade: A flashbang that additionally releases a spray of rubber pellets.

Syndicate:
- Cluster grenade(8TC): Splits into three explosive grenades after a 3.5 second timer, these grenades explode one second after splitting up.
- Shrapnel grenade(4TC): A grenade that instead of exploding releases a spray of lethal bullets that pierce mobs in an area around itself.
- Incendiary grenade(4TC): A shrapnel grenade that shoots piercing incendiary bullets, igniting anyone near the detonation area.
- Cluster banana peel(clown exclusive, 6TC): After landing on the floor it splits into 6 explosive banana peels, these banana peels have a 20 second timer if nobody slips on them.
- Slipocalypse ClusterSoap(3TC): After landing on the floor it litters the area with 30 small pieces of syndicate soap, these pieces of soap evaporate after 60 seconds.

Other changes:

- Buffed the explosive banana peel so that it deals 60 damage to an unarmored target when stepped on.

**Media**

https://github.com/space-wizards/space-station-14/assets/137322659/aae3de4b-660b-4a9e-a632-047e0577fd54

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Dygon
- add: A new type of grenade has been developed, these grenades release a spray of bullets instead of exploding: Shrapnel grenade, Stinger grenade and Incendiary grenade.
- add: Stinger grenades and ClusterBangs are now available in the sectech.
- add: Incendiary grenades, Shrapnel grenades, Cluster grenades and the Slipocalypse ClusterSoap are now available in the syndicate uplink, clowns also get a Clusterbanana in their uplink.
- tweak: Explosive banana peels are now more dangerous.

### CODE REVIEW

**Comment by UbaserB** [comment]
Insanely cool PR. It might be better to atomise this or greatly cut down on changelog length.

**Comment by UbaserB** [comment]
Also if its ok can i make a custom sprite for the incendiary grenade?

**Comment by Arendian** [comment]
Small issue it has at the moment that doesn't show in the video(video is from old pr) is that the detonation sounds don't work because when the entity gets deleted the EmitSoundOnTrigger gets cut off if the entity gets deleted before the sound has ended.
 I have no idea if this is because of the current sound bugs or because of a permanent change making EmitSoundOnTrigger and DeleteOnTrigger not compatible with eachother.
 If it's a permanent change I will need to find a solution for it because at the moment the flashbang also doesn't make any detonatiion sound on live.

**Comment by Arendian** [comment]
> Also if its ok can i make a custom sprite for the incendiary grenade?

Gladly!, I am pretty bad at spriting and just made them recoloured explosive grenades :sweat_smile:

**Comment by Arendian** [comment]
> Insanely cool PR. It might be better to atomise this or greatly cut down on changelog length.

I can probably generalize the added grenades instead of explaining what they do individually

**Comment by JoeHammad1844** [comment]
I like the stinger grenades and how officers will need to put more thought into throwing them, and will be really funny when recursive explosions blows it up on them and they take all the pellets at once

File: Content.Server/Explosion/EntitySystems/ClusterGrenadeSystem.cs [hunk 1]
```diff
@@ -58,59 +60,86 @@ private void OnClugUsing(Entity<ClusterGrenadeComponent> clug, ref InteractUsing
...
-        args.Handled = true;
     }
 
-    private bool TryGetGrenade(Entity<ClusterGrenadeComponent> ent, out EntityUid grenade)
+    private void ThrowGrenade(EntityUid grenade, Angle angle, ClusterGrenadeComponent clug)
+    {
+        if (clug.RandomSpread)
+            _throwingSystem.TryThrow(grenade, angle.ToVec().Normalized() * _random.NextFloat(clug.MinSpreadDistance, clug.MaxSpreadDistance), clug.Velocity);
+        else _throwingSystem.TryThrow(grenade, angle.ToVec().Normalized() * clug.Distance, clug.Velocity);
+    }
```
Comment (EmoGarbage404) [inline Content.Server/Explosion/EntitySystems/ClusterGrenadeSystem.cs:133]: the formatting on both of these is kinda weird. you could just do an angle variable with a ternary operator and it'd look a lot cleaner.

File: Content.Server/Explosion/EntitySystems/ClusterGrenadeSystem.cs [hunk 2]
```diff
@@ -58,59 +60,86 @@ private void OnClugUsing(Entity<ClusterGrenadeComponent> clug, ref InteractUsing
...
+    private bool TryGetGrenade(EntityUid clugUid, ClusterGrenadeComponent component, out EntityUid grenade)
     {
         grenade = default;
-        var component = ent.Comp;
 
         if (component.UnspawnedCount > 0)
         {
             component.UnspawnedCount--;
-            grenade = EntityManager.SpawnEntity(component.FillPrototype, Transform(ent).MapPosition);
+            grenade = EntityManager.SpawnEntity(component.FillPrototype, Transform(clugUid).MapPosition);
```
Comment (EmoGarbage404) [inline Content.Server/Explosion/EntitySystems/ClusterGrenadeSystem.cs:142]: not your code but replace EntityManager.SpawnEntity with Spawn()

File: Resources/Prototypes/Catalog/VendingMachines/Inventories/sec.yml [hunk 3]
```diff
@@ -3,6 +3,8 @@
   startingInventory:
     Handcuffs: 8
     GrenadeFlashBang: 4
+    ClusterBangFull: 4
+    GrenadeStinger: 8
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Catalog/VendingMachines/Inventories/sec.yml:7]: these could be toned down a little bit. there's quite a few in here.

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]

File: Content.Server/Explosion/Components/ClusterGrenadeComponent.cs [hunk 4]
```diff
@@ -27,17 +27,81 @@ public sealed partial class ClusterGrenadeComponent : Component
         [DataField("maxGrenadesCount")]
         public int MaxGrenades = 3;
 
+        // <summary>
+        ///     Maximum delay in milliseconds between grenade triggers
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (Arendian) [reply Content.Server/Explosion/Components/ClusterGrenadeComponent.cs]: I forgot to change the comment after changing from milliseconds to seconds

File: Resources/Prototypes/Catalog/VendingMachines/Inventories/sec.yml (hunk 3 above)
Comment (Arendian) [reply Resources/Prototypes/Catalog/VendingMachines/Inventories/sec.yml]: Halved both

File: Content.Server/Explosion/Components/ClusterGrenadeComponent.cs [hunk 5]
```diff
@@ -27,17 +27,81 @@ public sealed partial class ClusterGrenadeComponent : Component
...
+        [ViewVariables(VVAccess.ReadWrite)]
+        [DataField("bombletDelayMin")]
+        public float BombletDelayMin = 0f;
+
+        // <summary>
+        ///     Minimum delay in milliseconds before bomblets start to be triggered.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        [DataField("minimumDelay")]
+        public float MinimumDelay = 1.0f;
```
[the remark this answers is not in the document]
Comment (Arendian) [reply Content.Server/Explosion/Components/ClusterGrenadeComponent.cs]: Changed all bomblet references to grenade, is current name/description clear enough?

File: Content.Server/Explosion/EntitySystems/ClusterGrenadeSystem.cs (hunk 1 above)
Comment (Arendian) [reply Content.Server/Explosion/EntitySystems/ClusterGrenadeSystem.cs]: Changed the formatting

**Comment by Arendian** [comment]
Removed the new china lake ammo and will make a different PR for it

**Review by EmoGarbage404 — APPROVED** [review]

**Comment by JoeHammad1844** [comment]
oh boy I cant wait to ignite the entire emergency shuttle on fire because the courser is so small!

**Pull request merged**